### DISCUSSION TITLE
[question]: supporting row level security (RLS) in Postgres SQL

### DISCUSSION THREAD

**Discussion by charlie-maloney** [mikro-orm/mikro-orm#6137] [Q&A]
Hey! I am looking into adopting Mikro in a new project, and wanted to ask about how I could support row level security in Postgres. 

Essentially, for each request (or connection in a request's lifecycle), I want to set dynamic configuration parameters (e.g. `tenant_id`) via the `set_config` Postgres function. 

For example, I could use the `em.getConnection()` function to get an explicit connection, set the proper variables, and perform a query:

```ts
    const conn = this.em.getConnection();

    await conn.execute("SELECT set_config('app.bypass_rls', 'off', FALSE)");

    await conn.execute(
      `SELECT set_config('app.current_tenant', '${tenantId}', FALSE)`,
    );

    await conn.execute(
      `SELECT * from rls_protected_table`,
    );

```

However, since a lot of the magic behind Mikro abstracts away the connection logic, I am wondering what the best way to ensure that `app.current_tenant` is properly set on each connection. For example, when a query for a entity using a repository or call `em.flush()`, I want to ensure the correct tenant is being set on the connection.  

I have explored a few approaches, but not with much luck:
* use a different ORM instance for each tenant (though, I believe this means that I will create a new pool for each ORM instance)
* hook into connection lifecycle events via knex / tarn (seems quite hacky)

Has anyone implemented anything similar? Or perhaps have suggestions?

**Comment by B4nan** [comment]
> use a different ORM instance for each tenant (though, I believe this means that I will create a new pool for each ORM instance)

You can reuse one knex instance via `driverOptions` (see https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/reusing-knex-client.test.ts), but it feels more hacky than using the knex connection hooks, which I would suggest exploring first (in combination with async local storage I guess).

Another way could be using explicit transactions, that way all the queries will use the same connection (the transaction context). That feels more "native" for sure, but also means you will have transactions around everything and there will be an implicit flush call.

**Comment by charlie-maloney** [comment]
Thanks for the quick response @B4nan! 

I didn't realize you could use one pool -- I'll look into that. Sounds like there are some options, but just wanted to make sure I wasn't trying to solve a problem that has already been solved before I did more digging. 

Also, is this the documentation I should refer to regarding the explicit transactions you are talking about? 
https://mikro-orm.io/docs/transactions#approach-2-explicitly

**Reply by B4nan** [reply]
Yeah, I can imagine you would have a custom middleware for the request context, and there you would start a transaction via `em.begin()` and set the tenant on that connection. That is currently the only way to maintain a single connection. I might look into native support for RLS at some point, maybe v7, maybe sooner, but it will require some non trivial refactoring (which would happen in v7 anyway most likely).

**Reply by charlie-maloney** [reply]
Sounds good. I will weigh the options and see which works best for me. 

Appreciate your work!

**Reply by robbyemmert** [reply]
@B4nan I have a bunch of existing code that uses implicit (and maybe a few explicit) transactions. We are currently looking for a way to switch to a multi tenant system via RLS. Would you recommend simply using disableTransactions with the above solution, or would hooking into EntityManager/knex in some way be cleaner?

Thank you for your work, by the way. We’re loving 6.x. When we upgraded I was impressed with the performance upgrades and improved docs. I know there is some pretty complex work under the hood to make that happen. 

Let me know if any real world implementation examples would be helpful to your v7 efforts. We’re planning a multi tenant implementation for Q1 next year.

**Reply by B4nan** [reply]
`disableTransactions` won't help here, with that option there won't be any connection handle on the EM fork, and `em.begin` and all the other tx methods with this flag enabled are noop. 

I would try the knex connection hooks personally, with a custom ALS to keep your tenant id at hand.

> For example, I could use the em.getConnection() function to get an explicit connection, set the proper variables, and perform a query:

Also I just realized I haven't commented on this part of the OP. This won't work either, `em.getConnection()` won't give you explicit connection, it gives you a singleton connection wrapper, which currently represents a knex instance. Calling this method multiple times will always give you the same result (with the exception of using read replicas, there one replica means one connection instance).

MikroORM currently only maintains the knex instance, there is no connection handling on our level, its all delegated.

**Reply by albertfdp** [reply]
Hi, I have tried to follow @B4nan recommendation on my project, where I use Nestjs + Mikro ORM and CLS. I have implemented an interceptor that looks like this:

```ts
import { EntityManager } from "@mikro-orm/postgresql";
import { CallHandler, ExecutionContext, Injectable, NestInterceptor } from "@nestjs/common";
import { ClsServiceManager } from "nestjs-cls";
import { catchError, map, Observable, throwError } from "rxjs";

@Injectable()
export class TenantInterceptor implements NestInterceptor {
  constructor(private readonly em: EntityManager) {}

  async intercept(context: ExecutionContext, next: CallHandler<any>): Promise<Observable<any>> {
    const cls = ClsServiceManager.getClsService();
    const auth = cls.get("auth");

    await this.em.begin();

    if (auth && auth.orgId) {
      const connection = this.em.getConnection();
      await connection.execute(`SELECT set_config('app.current_org_id', '${auth.orgId}', TRUE);`);
    }

    return next.handle().pipe(
      map(async data => {
        await this.em.commit();
        return data;
      }),
      catchError(async error => {
        await this.em.rollback();
        return throwError(() => error);
      }),
    );
  }
}
```

However, I have seen that this only works when `SELECT set_config('app.current_org_id', '${auth.orgId}', TRUE);` is set to `TRUE`, while I would like it to be `FALSE` as that ensures that this is only set for the current transaction ([source](https://pgpedia.info/s/set_config.html)). I am curious to know if there is an alternative to run a raw SQL query that ensures that only runs in the same transaction. 

I am also using this in combination with the following middleware

```ts
app.use((request: Request, response: Response, next: NextFunction) => {
    const orm = app.get<MikroORM>(MikroORM);

    RequestContext.create(orm.em, next);
  });
  ```

Really looking forward for RLS being part of v7. Thanks for the great work!

**Reply by B4nan** [reply]
>       const connection = this.em.getConnection();
>      await connection.execute(`SELECT set_config('app.current_org_id', '${auth.orgId}', TRUE);`);

This is wrong, you are not respecting the transaction context this way. Use `em.execute()` instead.

Not sure if your problem is about that, but without that you are executing on a new connection handle, not the one the transaction is using.

**Reply by albertfdp** [reply]
Oh...🤦 That was it then! That seems to work quite well now. 

This thread was very useful!

**Reply by albertfdp** [reply]
Leaving it here updated if anyone finds it useful in the future:

```ts
import { EntityManager } from "@mikro-orm/postgresql";
import { CallHandler, ExecutionContext, Injectable, NestInterceptor } from "@nestjs/common";
import { ClsServiceManager } from "nestjs-cls";
import { catchError, map, Observable, throwError } from "rxjs";

@Injectable()
export class TenantInterceptor implements NestInterceptor {
  constructor(private readonly em: EntityManager) {}

  async intercept(context: ExecutionContext, next: CallHandler<any>): Promise<Observable<any>> {
    const cls = ClsServiceManager.getClsService();
    const auth = cls.get("auth");

    await this.em.begin();

    if (auth && auth.orgId) {
      await this.em.execute(`SELECT set_config('app.current_org_id', '${auth.orgId}', TRUE);`);
    }

    return next.handle().pipe(
      map(async data => {
        await this.em.commit();
        return data;
      }),
      catchError(async error => {
        await this.em.rollback();
        return throwError(() => error);
      }),
    );
  }
}
```

**Reply by leonardbinet** [reply]
👍 just a small note, wrapping all operations resulting from a request in a single transaction can have performance implications if some http request or job results in multiple sql requests operating on large numbers of records (this might end up in deadlocks). In some case you might want **not** to wrap everything in a single transaction.

**Reply by robbyemmert** [reply]
@B4nan does the new `onQuery` setting in Configuration open the door to setting RLS parameters before each query? Are there any nuances with migrations, etc we need to be aware of, or can we use this to set query parameters such as default schema, or security parameters?

Example (set default schema):
```ts
onQuery: (sql, params) => {
    const schema = connection.schema ?? 'public'

    return `SET search_path TO ${schema};\n${sql}`
}
 ```

**Reply by B4nan** [reply]
Not sure if this would work for transactions, since the `begin/commit/rollback` queries are not executed through the ORM like the rest of the queries. But for reading purposes, yeah, I guess this could work. For writes, I'd rather suggest using transactions, and calling this once after the transaction is opened, since that will ensure the same connection is used behind the scenes.

**Reply by sherif-ubiheal** [reply]
>  I might look into native support for RLS at some point, maybe v7, maybe sooner

@B4nan, is v7 the target for that?

**Reply by B4nan** [reply]
No, not really, I got way too many other things, and I don't want to postpone v7 forever. Best way to see this happen is to contribute.

**Reply by leonardbinet** [reply]
@B4nan here are a few thoughts on what such feature could look like.

In my opinion it should include:
- a way to declare policies, that will be attached to tables (will impact generated migrations). Here is some [inspiration from drizzle on this topic](https://orm.drizzle.team/docs/rls#policies)
- a way to specify a parameter on which isolation will be based (for instance specify the tenant you want to be scoped on)

There are several ways to pass information:
- in postgres username, role, or in connection params, all of which require a dedicated connection per tenant (which can quickly become too expensive in terms of connections)
- via local variables: `SET LOCAL tenant_id = xxx`, which will be effective during the transaction. Such approach is much more scalable but requires to wrap all statements in transactions.

Given the ORM philosophy with unit of work, one option I think suitable for local variables, is to add to `entityManager.fork` method the ability to [specify such variables](https://www.postgresql.org/docs/current/sql-set.html).

Doing so would wrap this entity manager fork operations in transactions, which would redeclare for each transaction the passed variables.
One point to keep in mind is that it wouldn't make sense for a single entity manager to perform multiple requests with different "variables' context", otherwise the identity map might be populated with entities accessible under disparate conditions (for instances entities of different tenants, which couldn't be flushed together).


On our end, to implement RLS:
- it was quite easy to manually add policies in our migrations
- but it seemed out of reach to wrap all operations in transactions with the variable, so we ended up creating a dedicated connection per tenant (which isn't scalable enough), so that would be the primary need to have a real RLS feature

I think contributing to the policy declaration, or working on a guide is something in my reach, on the other hands, giving the ability to fork the entity manager with local variables, and wrap all operations in transactions with those variables might be too big of a challenge given my (absence of) knowledge of the internals of mikro-orm.

**Reply by B4nan** [reply]
So the ideal solution would allow using multiple connections without a transaction, but once the tenant variable is set on the connection level, it should be only used for that tenant? Or is it fine if we override the tenant variable when acquiring the connection from the pool? 

The transaction context the ORM holds is nothing more than a connection handle (low level, knex/kysely connection), we could most likely use the same mechanism even without a transaction, but it would still mean a single connection per tenant, so likely the same scalability problem. If we delegate this to the connection pool, it might work too, without caring about the connection handles, we'd just have to make sure the acquired connection is correctly set up.

**Reply by leonardbinet** [reply]
Ah indeed, good point, I didn't think about that.
`Or is it fine if we override the tenant variable when acquiring the connection from the pool?` -> this is definitely the way to to go, so that the number of connections doesn't grow indefinitely (it would if connections are reserved for a single tenant).

With such mechanism, where the session variable is set when the connection is acquired from the pool, and reset when the connection is released, there would be no need for wrapping transactions 👍

**Reply by leonardbinet** [reply]
to wrap it up, I'll correct my initial message.

There are several ways to pass information:

- in postgres username, role, or in connection params, all of which require a dedicated connection per tenant (which can quickly become too expensive in terms of connections)
- via local variables: `SET LOCAL tenant_id = 12`, which will be effective during the transaction. Such approach is much more scalable but requires to wrap all statements in transactions.
- (new) via session variable: `SET SESSION tenant_id = 12`. To be able to reuse existing connections, such mechanism would require to:
   -  set the variable when the connection is acquired from the pool `SET SESSION tenant_id = 12` (during `pool.acquire()`)
   -  reset the variable when the connection is released to the pool: `SET SESSION tenant_id = null` (during `pool.release()`)

**Reply by B4nan** [reply]
Ok, that might be actually doable already, connection pooling is implemented on knex level via https://github.com/vincit/tarn.js, I think there are some hooks for this. You could deal with the context via `AsyncLocalStorage` so you have a place where you can store the current tenant id.

(with v7 and kysely this is only used for mssql, it might be trickier there, but still likely doable the same way)

**Reply by leonardbinet** [reply]
Awesome, I'll try it when I find some time, I'll let you know of how it goes.

**Comment by ghempton** [comment]
Chiming in from the v7 + Kysely side — the ideal per-acquire pattern @leonardbinet described is actually already supported by Kysely's Postgres dialect via `onReserveConnection`, which fires on every `pool.connect()` and is awaited before the connection is handed back:

```js
// kysely/dist/esm/dialect/postgres/postgres-driver.js
async acquireConnection() {
  const client = await this.#pool.connect();
  // ... onCreateConnection (one-shot) ...
  if (this.#config.onReserveConnection) {
    await this.#config.onReserveConnection(connection); // per-acquire, awaited
  }
  return connection;
}
```

So for RLS we'd do:

```ts
onReserveConnection: async (connection) => {
  const { userId } = getActionContext(); // AsyncLocalStorage
  const sql = userId && isUuid(userId)
    ? `SET app.current_user_id = '${userId}'`
    : `RESET app.current_user_id`;
  await connection.executeQuery({ sql, parameters: [], query: {} as any, queryId: {} as any });
}
```

Because it runs on every acquire, stale values from a prior checkout are always overwritten before the next query runs — no cross-request leakage, no transaction required, works for auto-commit reads and flush-wrapped writes alike. Policies just need `NULLIF(current_setting('app.current_user_id', true), '')::uuid` to handle the `RESET → empty string` cast case.

The one snag: mikro-orm's `PostgreSqlConnection.createKyselyDialect` only forwards `onCreateConnection` to the dialect, not `onReserveConnection`. To use it today we have to subclass `PostgreSqlConnection` + `PostgreSqlDriver` to thread it through — small but awkward.

@B4nan — two questions:
1. Would a PR threading `onReserveConnection` through `Configuration` (alongside `onCreateConnection`) be welcome?
2. Does this overall approach (per-acquire session SET via `onReserveConnection`, ALS-driven user id, `NULLIF` in policies for the reset case) seem sound to you, or are there gotchas you'd flag from having thought about RLS for mikro-orm already?

**Reply by B4nan** [reply]
Sure, PR welcome, please target the `next` branch. No opinions on this, really. I haven't used RLS myself, but I'd like to take a closer look at native support in the near future.

**Reply by simonbrunel** [reply]
@ghempton Do you still plan to submit a PR to expose `onReserveConnection`? 

I'm considering adopting MikroORM for a platform that requires strict tenant isolation, and I'm wondering whether your approach is robust and guarantees that no data leakage can occur due to connection reuse (or missing tenant id)?

The other alternative would be Drizzle, which has [built-in support for RLS](https://orm.drizzle.team/docs/rls), but I feel like I’d lose too much switching ORMs.

**Reply by ghempton** [reply]
I haven't had time to submit the PR yet, going to be added RLS to our app in the coming week or two and will introduce the PR then.

**Reply by simonbrunel** [reply]
@ghempton Thank you so much for #7815, realizing just now that this has been released 2 weeks ago (thanks @B4nan)

**Comment by B4nan** [comment]
v7.2 will provide native support for RLS, see #7984
